File: qqbot-protocol/src/main/java/com/mieai/qqbot/protocol/gateway/GatewayIdentify.java
package com.mieai.qqbot.protocol.gateway;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.util.Map;
import java.util.Objects;

/** Data sent with an opcode 2 Identify command. */
@JsonIgnoreProperties(ignoreUnknown = true)
public record GatewayIdentify(
        String token, long intents, int[] shard, Map<String, String> properties) {
    public GatewayIdentify {
        Objects.requireNonNull(token, "token must not be null");
        Objects.requireNonNull(shard, "shard must not be null");
        Objects.requireNonNull(properties, "properties must not be null");
        shard = shard.clone();
        properties = Map.copyOf(properties);
    }

    @Override
    public int[] shard() {
        return shard.clone();
    }

    @Override
    public String toString() {
        return "GatewayIdentify[token=<redacted>, intents=" + intents + ", shard=<redacted>]";
    }
}

File: qqbot-protocol/src/main/kotlin/com/mieai/qqbot/protocol/gateway/GatewayIdentify.kt
package com.mieai.qqbot.protocol.gateway

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty

/** Data sent with an opcode 2 Identify command. */
@JsonIgnoreProperties(ignoreUnknown = true)
class GatewayIdentify @JsonCreator(mode = JsonCreator.Mode.PROPERTIES) constructor(
    @param:JsonProperty("token")
    @get:JsonProperty("token")
    val token: String,
    @param:JsonProperty("intents")
    @get:JsonProperty("intents")
    val intents: Long,
    @param:JsonProperty("shard") shard: IntArray,
    @param:JsonProperty("properties") properties: Map<String, String>,
) {
    private val shardValue = shard.clone()

    @get:JsonProperty("shard")
    val shard: IntArray
        get() = shardValue.clone()

    @get:JsonProperty("properties")
    val properties: Map<String, String> = properties.toMap()

    override fun toString(): String = "GatewayIdentify(token=<redacted>, intents=$intents, shard=<redacted>)"
}
